modules/stochastic_tools/src/surrogates/BasicNNTrainer.C

```diff
@@ -0,0 +1,185 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "BasicNNTrainer.h"
+#include "Sampler.h"
+
+registerMooseObject("StochasticToolsApp", BasicNNTrainer);
+
+InputParameters
+BasicNNTrainer::validParams()
+{
+  InputParameters params = SurrogateTrainer::validParams();
+
+  params.addClassDescription("Trains a simple neural network using libtorch.");
+
+  MooseEnum data_type("real=0 vector_real=1", "real");
+  params.addRequiredParam<ReporterName>(
+      "response",
+      "Reporter value of response results, can be vpp with <vpp_name>/<vector_name> or sampler "
+      "column with 'sampler/col_<index>'.");
+  params.addParam<MooseEnum>("response_type", data_type, "Response data type.");
+  params.addParam<std::vector<ReporterName>>(
+      "predictors",
+      std::vector<ReporterName>(),
+      "Reporter values used as the independent random variables, If 'predictors' and "
+      "'predictor_cols' are both empty, all sampler columns are used.");
+  params.addParam<std::vector<unsigned int>>(
+      "predictor_cols",
+      std::vector<unsigned int>(),
+      "Sampler columns used as the independent random variables, If 'predictors' and "
+      "'predictor_cols' are both empty, all sampler columns are used.");
+  params.addParam<unsigned int>("no_batches", 1, "Number of batches.");
+  params.addParam<unsigned int>("no_epochs", 1, "Number of epochs.");
+  params.addParam<unsigned int>("no_hidden_layers", 0, "Number of hidden layers.");
+  params.addParam<std::vector<unsigned int>>(
+      "no_neurons_per_layer", std::vector<unsigned int>(), "Number of neurons per layer.");
+  params.addParam<Real>("learning_rate", 0.001, "Learning rate (relaxation).");
+
+  return params;
+}
+
+BasicNNTrainer::BasicNNTrainer(const InputParameters & parameters)
+  : SurrogateTrainer(parameters),
+    _sampler_row(getSamplerData()),
+    _sample_points(declareModelData<std::vector<std::vector<Real>>>("_sample_points")),
+    _pvals(getParam<std::vector<ReporterName>>("predictors").size()),
+    _pcols(getParam<std::vector<unsigned int>>("predictor_cols")),
+    _response(getTrainingData<Real>(getParam<ReporterName>("response"))),
+    _no_batches(getParam<unsigned int>("no_batches")),
+    _no_epocs(getParam<unsigned int>("no_epochs")),
+    _no_hidden_layers(getParam<unsigned int>("no_hidden_layers")),
+    _no_neurons_per_layer(getParam<std::vector<unsigned int>>("no_neurons_per_layer")),
+    _learning_rate(getParam<Real>("learning_rate"))
+{
+  const auto & pnames = getParam<std::vector<ReporterName>>("predictors");
```

Comment:
```suggestion
#ifndef TORCH_ENABLED
  mooseError("PyTorch C++ API (libtorch) must be installed to use ", type(), ", see https://mooseframework.inl.gov/modules/stochastic_tools/install_pytorch.html for instruction.");
#endif

  const auto & pnames = getParam<std::vector<ReporterName>>("predictors");
```

Reply:
Very nice error message, thanks Zach! I added it to every similar case!